File: packages/sdk-node/src/phase3.ts
// ============================================================
// @kangopenbanking/sdk — Phase 3 helpers
// Additive: introduced in SDK v1.4.0. Does NOT modify any
// existing method, type, or resource. Safe to tree-shake.
// Standards cited: PSD2 RTS Art. 36 (event integrity), Stripe
// API Reference (exports & webhook reliability baseline).
// ============================================================

import type { KangOpenBanking } from "./client";

export type ExportFormat = "csv" | "xlsx" | "json";

export interface ExportFilters {
  merchant_id: string;
  from?: string;            // ISO date (YYYY-MM-DD)
  to?: string;              // ISO date
  environment?: "sandbox" | "live";
  currency?: string;        // XAF, XOF, USD, EUR, NGN, ...
  format?: ExportFormat;
}

export interface ExportJob {
  export_id: string;
  status: "queued" | "processing" | "ready" | "failed";
  format: ExportFormat;
  download_url?: string;
  expires_at?: string;
}

export interface StatementRequest {
  merchant_id: string;
  month: string;            // YYYY-MM
  format?: "json" | "csv" | "pdf";
}

export interface ReconciliationRunRequest {
  merchant_id: string;
  from: string;
  to: string;
}

export interface MerchantApiKey {
  id: string;
  prefix: string;
  scopes: string[];
  environment: "sandbox" | "live";
  created_at: string;
  last_used_at?: string | null;
}

export interface MerchantWebhookEndpoint {
  id: string;
  url: string;
  events: string[];
  is_active: boolean;
  created_at: string;
}

export interface WebhookDelivery {
  id: string;
  endpoint_id: string;
  event_type: string;
  status: "pending" | "delivered" | "failed";
  attempt: number;
  max_attempts: number;
  response_status: number | null;
  created_at: string;
}

/**
 * Phase-3 merchant operations: exports, statements, reconciliation,
 * API key lifecycle and webhook reliability.
 *
 * Wire it onto an existing KangOpenBanking instance:
 *   const kob = new KangOpenBanking({...});
 *   const merchant = new MerchantOpsResource(kob);
 *   await merchant.exports.transactions({ merchant_id, from, to });
 */
export class MerchantOpsResource {
  constructor(private client: KangOpenBanking) {}

  readonly exports = {
    transactions: (filters: ExportFilters): Promise<ExportJob> =>
      this.client.request("POST", "merchant-exports", { resource: "transactions", ...filters }),
    settlements: (filters: ExportFilters): Promise<ExportJob> =>
      this.client.request("POST", "merchant-exports", { resource: "settlements", ...filters }),
    fees: (filters: ExportFilters): Promise<ExportJob> =>
      this.client.request("POST", "merchant-exports", { resource: "fees", ...filters }),
    get: (exportId: string): Promise<ExportJob> =>
      this.client.request("GET", "merchant-exports", undefined, { export_id: exportId }),
  };

  readonly statements = {
    download: (req: StatementRequest): Promise<{ url: string; expires_at: string }> =>
      this.client.request("GET", "gateway-merchant-statement", undefined, {
        merchant_id: req.merchant_id,
        month: req.month,
        format: req.format ?? "pdf",
      }),
  };

  readonly reconciliation = {
    run: (req: ReconciliationRunRequest): Promise<{ run_id: string; status: string }> =>
      this.client.request("POST", "gateway-reconciliation-run", req),
    get: (runId: string): Promise<{ run_id: string; status: string; mismatches: number }> =>
      this.client.request("GET", "gateway-reconciliation-run", undefined, { run_id: runId }),
  };

  readonly apiKeys = {
    list: (merchantId: string): Promise<MerchantApiKey[]> =>
      this.client.request("GET", "gateway-merchant-api-keys", undefined, { merchant_id: merchantId }),
    create: (params: { merchant_id: string; label: string; scopes: string[]; environment: "sandbox" | "live" }): Promise<MerchantApiKey & { plaintext_key: string }> =>
      this.client.request("POST", "gateway-merchant-api-keys", { action: "create", ...params }),
    revoke: (keyId: string): Promise<{ id: string; revoked: true }> =>
      this.client.request("POST", "gateway-merchant-api-keys", { action: "revoke", key_id: keyId }),
    rotate: (keyId: string): Promise<MerchantApiKey & { plaintext_key: string }> =>
      this.client.request("POST", "gateway-merchant-api-keys", { action: "rotate", key_id: keyId }),
  };

  readonly webhooks = {
    listEndpoints: (merchantId: string): Promise<MerchantWebhookEndpoint[]> =>
      this.client.request("GET", "gateway-webhook-endpoints", undefined, { merchant_id: merchantId }),
    listDeliveries: (params: { endpoint_id: string; limit?: number; status?: string }): Promise<WebhookDelivery[]> =>
      this.client.request("GET", "gateway-webhook-deliveries", undefined, {
        endpoint_id: params.endpoint_id,
        ...(params.limit ? { limit: String(params.limit) } : {}),
        ...(params.status ? { status: params.status } : {}),
      }),
    replayDelivery: (params: { endpoint_id: string; delivery_id: string }): Promise<{ replay_delivery_id: string; status: string }> =>
      this.client.request("POST", "gateway-webhook-replay-delivery", params),
    rotateSecret: (endpointId: string): Promise<{ id: string; secret: string; rotated_at: string }> =>
      this.client.request("POST", "gateway-webhook-endpoints", { action: "rotate_secret", endpoint_id: endpointId }),
  };
}

File: packages/sdk-python/kangopenbanking/phase3.py
"""
Phase 3 helpers — additive merchant operations module.

Introduced in kangopenbanking 1.4.0. Does NOT modify any existing
client method or type. Wire onto an existing client instance:

    from kangopenbanking import KangOpenBanking
    from kangopenbanking.phase3 import MerchantOps

    kob = KangOpenBanking(client_id="...", api_key="sbx_...")
    ops = MerchantOps(kob)
    job = ops.exports_transactions(merchant_id="mch_...", from_date="2026-04-01")

Standards cited: PSD2 RTS Art. 36, Stripe API Reference (Reports & Webhooks).
"""

from __future__ import annotations
from typing import Any, Dict, List, Literal, Optional


ExportFormat = Literal["csv", "xlsx", "json"]
ResourceKind = Literal["transactions", "settlements", "fees"]


class MerchantOps:
    """Phase 3 merchant operations — exports, statements, reconciliation,
    API keys, webhook deliveries."""

    def __init__(self, client: Any) -> None:
        self._c = client

    # ─── Exports ────────────────────────────────────────────────────
    def _export(self, resource: ResourceKind, **filters: Any) -> Dict[str, Any]:
        body = {"resource": resource}
        body.update({k: v for k, v in filters.items() if v is not None})
        return self._c._request("POST", "merchant-exports", json=body)

    def exports_transactions(
        self,
        merchant_id: str,
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
        environment: Optional[Literal["sandbox", "live"]] = None,
        currency: Optional[str] = None,
        format: ExportFormat = "csv",
    ) -> Dict[str, Any]:
        return self._export("transactions",
                            merchant_id=merchant_id, **{"from": from_date, "to": to_date},
                            environment=environment, currency=currency, format=format)

    def exports_settlements(self, merchant_id: str, **filters: Any) -> Dict[str, Any]:
        return self._export("settlements", merchant_id=merchant_id, **filters)

    def exports_fees(self, merchant_id: str, **filters: Any) -> Dict[str, Any]:
        return self._export("fees", merchant_id=merchant_id, **filters)

    def export_get(self, export_id: str) -> Dict[str, Any]:
        return self._c._request("GET", "merchant-exports", params={"export_id": export_id})

    # ─── Statements ─────────────────────────────────────────────────
    def statement_download(self, merchant_id: str, month: str, format: str = "pdf") -> Dict[str, Any]:
        return self._c._request("GET", "gateway-merchant-statement",
                                params={"merchant_id": merchant_id, "month": month, "format": format})

    # ─── Reconciliation ─────────────────────────────────────────────
    def reconciliation_run(self, merchant_id: str, from_date: str, to_date: str) -> Dict[str, Any]:
        return self._c._request("POST", "gateway-reconciliation-run",
                                json={"merchant_id": merchant_id, "from": from_date, "to": to_date})

    def reconciliation_get(self, run_id: str) -> Dict[str, Any]:
        return self._c._request("GET", "gateway-reconciliation-run", params={"run_id": run_id})

    # ─── Merchant API Keys ──────────────────────────────────────────
    def api_keys_list(self, merchant_id: str) -> List[Dict[str, Any]]:
        data = self._c._request("GET", "gateway-merchant-api-keys",
                                params={"merchant_id": merchant_id})
        return data if isinstance(data, list) else data.get("data", [])

    def api_key_create(self, merchant_id: str, label: str, scopes: List[str],
                        environment: Literal["sandbox", "live"]) -> Dict[str, Any]:
        return self._c._request("POST", "gateway-merchant-api-keys", json={
            "action": "create", "merchant_id": merchant_id,
            "label": label, "scopes": scopes, "environment": environment,
        })

    def api_key_revoke(self, key_id: str) -> Dict[str, Any]:
        return self._c._request("POST", "gateway-merchant-api-keys",
                                json={"action": "revoke", "key_id": key_id})

    def api_key_rotate(self, key_id: str) -> Dict[str, Any]:
        return self._c._request("POST", "gateway-merchant-api-keys",
                                json={"action": "rotate", "key_id": key_id})

    # ─── Merchant Webhooks ──────────────────────────────────────────
    def webhook_endpoints(self, merchant_id: str) -> List[Dict[str, Any]]:
        data = self._c._request("GET", "gateway-webhook-endpoints",
                                params={"merchant_id": merchant_id})
        return data if isinstance(data, list) else data.get("data", [])

    def webhook_deliveries(self, endpoint_id: str, limit: int = 50,
                           status: Optional[str] = None) -> List[Dict[str, Any]]:
        params = {"endpoint_id": endpoint_id, "limit": str(limit)}
        if status:
            params["status"] = status
        data = self._c._request("GET", "gateway-webhook-deliveries", params=params)
        return data if isinstance(data, list) else data.get("data", [])

    def webhook_replay(self, endpoint_id: str, delivery_id: str) -> Dict[str, Any]:
        return self._c._request("POST", "gateway-webhook-replay-delivery",
                                json={"endpoint_id": endpoint_id, "delivery_id": delivery_id})

    def webhook_rotate_secret(self, endpoint_id: str) -> Dict[str, Any]:
        return self._c._request("POST", "gateway-webhook-endpoints",
                                json={"action": "rotate_secret", "endpoint_id": endpoint_id})


__all__ = ["MerchantOps", "ExportFormat", "ResourceKind"]
